what4/src/What4/Protocol/SMTWriter.hs

```diff
@@ -787,6 +786,23 @@ cacheLookup
 cacheLookup conn lookup_action =
   readIORef (entryStack conn) >>= firstJustM lookup_action
 
+
+-- | Like 'findM', but also allows you to compute some additional information in the predicate.
+firstJustM :: Monad m => (a -> m (Maybe b)) -> [a] -> m (Maybe b)
+firstJustM _ [] = pure Nothing
+firstJustM p (x:xs) = maybeM (firstJustM p xs) (pure . Just) (p x)
```

Comment:
I normally use `return` in the `Monad` context and only use `pure` in `Applicative` contexts.